Lazy query execution support

svladykin opened the pull request:
Allows to lazily traverse large data sets without generating the whole result set into a separate place. 

Currently works only in embedded mode (network protocol relies on knowing a result set size) and has restrictions for having multiple active result sets from a single connection as well as for concurrent usage of the same connection between threads.

Please review.

svladykin committed:
Lazy query execution support

grandinj commented on h2/src/main/org/h2/engine/Session.java:
```diff
@@ -542,11 +554,8 @@ public Command prepareLocal(String sql) {
-                modificationMetaID = database.getModificationMetaId();
-            } else {
+        if (queryCache != null) {
+            synchronized (queryCache) {
```
why do you need to synchronize here? Surely at this point we either hold the Database object or the Session object locked already?

svladykin replied:
Yes, I double checked, you are right. This method should not be executed concurrently. I'll revert this part.

grandinj commented on h2/src/main/org/h2/command/Command.java:
```diff
@@ -44,7 +43,7 @@
     private final String sql;
 
-    private boolean canReuse;
+    private volatile boolean canReuse;
```
????

svladykin replied:
Will revert.

svladykin commented:
@thomasmueller @grandinj Can you please take a look?

grandinj reviewed (commented):
Looking at this, I wonder what the performance impact would be if we made the normal case just call the lazy inner classes - I'm mostly worried about duplicated code getting out of sync.

svladykin commented:
As for calling lazy instance for a usual case, you are right, it was done to avoid code duplication. I think it will not have any measurable performance impact (we generate lots of temporary objects anyways: Row, Value, Value[]...), but I can do some benchmarking if there are any doubts.

svladykin committed:
Merge branch 'master' of https://github.com/h2database/h2database into lazyQuery

svladykin committed:
Reverted synchronization from query cache

svladykin commented:
Reverted synchronization, any other notes?

svladykin commented:
@thomasmueller Could you please provide your opinion on this PR? 
We want to include this feature into upcoming Apache Ignite 2.0 and would prefer to start testing it early. Thus it would be nice to have a new version of H2 with this feature released asap.

thomasmueller commented:
Hi,

I'm on vacation this week and will not be able to help much. I wanted to
support lazy execution since a long time, and if possible always use it
(where possible), so this is a good feature. But I didn't have time to
review the patch yet. Of course this needs to be well rested, with ability
to disable for now. Also code duplication needs to be avoided. I will be
able to do a release in one week.

Regards
Thomas

Am 07.04.2017 1:29 nachm. schrieb "Sergi Vladykin" <notifications@github.com
>:

> @thomasmueller <https://github.com/thomasmueller> Could you please
> provide your opinion on this PR?
> We want to include this feature into upcoming Apache Ignite 2.0 and would
> prefer to start testing it early. Thus it would be nice to have a new
> version of H2 with this feature released asap.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2database/h2database/pull/488#issuecomment-292513179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA46GXcp-0GMc5YwL0f98TRSMfXNQwXzks5rth4tgaJpZM4Mw9hx>
> .
>

svladykin commented:
Good to know, thanks a lot!

svladykin commented:
@thomasmueller Did you have a chance to take a look?

thomasmueller commented:
Hi,

It is a large patch. I have to admit I didn't fully read it, but in general
it looks good to me. My feedback is mainly cosmetic:

> if (hasRecursive && p.isQuery() && p instanceof Query) {

instanceof should be avoided if possible. "hasRecursive" is parser wide,
but there might be multiple queries (subqueries) per SQL statement.
Would it be possible to call "setNeverLazy" while parsing recursive queries
/ while parsing "explain" queries?

> queryWithoutCache0

I kind of understand the reason for the name, but I don't agree it's good
name. Names can be long, that's fine, but using numbers is not good. It is
not clear when this method should be called versus the other. Also, not
sure if it needs to be protected.

Regards,
Thomas




On Sunday, April 16, 2017, Sergi Vladykin <notifications@github.com> wrote:

> @thomasmueller <https://github.com/thomasmueller> Did you have a chance
> to take a look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2database/h2database/pull/488#issuecomment-294342734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA46GSYWSm5FcutsQVHiY4zwrTnHD9YLks5rweE3gaJpZM4Mw9hx>
> .
>

svladykin committed:
queryWithoutCacheLazyCheck

svladykin committed:
Call Query.setNeverLazy only in Parser

svladykin committed:
minor

svladykin committed:
fix test

svladykin commented:
Good points!

I agree that it is preferable to call `setNeverLazy` only in needed places in `Parser`,  I've fixed that.
The only remaining place outside of `Parser` is `ViewIndex.findRecursive`, but this can not be fixed.

Renamed `queryWithoutCache0` to `	queryWithoutCacheLazyCheck` and made it private. Also moved some code from `SelectUnion` to `Query` to have everything in a single place.

Please look.

svladykin committed:
Minor code cleanup

svladykin committed:
Merge branch 'master' of https://github.com/h2database/h2database into lazyQuery

svladykin committed:
Fix TestQueryCache: for lazy queries it is important to close ResultSet

svladykin commented:
@thomasmueller Can you please take a look again? It would be nice to finish this PR and have a release this weekend.
